File: app/src/androidTest/java/com/shauvik/calc/JSONGen.java
package com.shauvik.calc;

import android.util.Log;
import android.util.Pair;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import org.junit.Test;

import java.lang.reflect.Modifier;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

class Model {
    public String name;
    public Map<String, ModelMethod> methods;
}

class ModelMethod {
    public String[] args;
    public String result;
}

public class JSONGen {
    @Test
    public void generate() {
        final Set<Model> models = genModels(com.shauvik.calc.model.Calculator.class);

        final Gson gson = new GsonBuilder().setPrettyPrinting().create();
        Log.d("MODEL OUTPUT", gson.toJson(models));
    }

    public Set<Model> genModels(Class target) {
        final HashSet<Model> models = new HashSet<>();
        genModels(models, target, new HashSet<>());
        return models;
    }

    public void genModels(Set<Model> models, Class target, Set<Class> visited) {
        if (visited.contains(target)) {
            return;
        }
        visited.add(target);

        final Model model = new Model();

        model.name = target.getSimpleName();

        model.methods = Arrays.stream(target.getDeclaredMethods())
                .filter((m) -> ((m.getModifiers() & Modifier.PUBLIC) != 0) && ((m.getModifiers() & Modifier.STATIC) == 0))
                .map((m) -> {
                    final ModelMethod method = new ModelMethod();

                    method.args = Arrays.stream(m.getParameterTypes())
                            .map(Class::getSimpleName)
                            .toArray(String[]::new);
                    method.result = m.getReturnType().getSimpleName();

                    genModels(models, m.getReturnType(), visited);

                    return new Pair<>(m.getName(), method);
                })
                .collect(HashMap::new, (m, method) -> m.put(method.first, method.second), HashMap::putAll);

        models.add(model);
    }
}

File: app/src/androidTest/java/com/shauvik/calc/JSONGen.kt
package com.shauvik.calc

import android.util.Log
import com.google.gson.GsonBuilder
import org.junit.Test
import java.lang.reflect.Modifier
import java.util.*

internal class Model {
    lateinit var methods: Map<String, ModelMethod>
}

internal class ModelMethod {
    lateinit var args: Array<String>
    lateinit var result: String
}

class JSONGen {
    @Test
    fun generate() {
        val models = genModels(com.shauvik.calc.model.Calculator::class.java)

        val gson = GsonBuilder().run {
            setPrettyPrinting()
            create()
        }
        Log.d("MODEL OUTPUT", gson.toJson(models))
    }

    internal fun genModels(target: Class<*>): Map<String, Model> {
        val models = HashMap<String, Model>()
        genModels(models, target, HashSet())
        return models
    }

    internal fun genModels(models: MutableMap<String, Model>, target: Class<*>, visited: MutableSet<Class<*>>) {
        if (visited.contains(target)) {
            return
        }
        visited.add(target)

        val model = Model().apply {
            methods = Arrays.stream(target.declaredMethods)
                    .filter { it.modifiers and Modifier.PUBLIC != 0 }
                    .filter { it.modifiers and Modifier.STATIC == 0 }
                    .map {
                        val method = ModelMethod().apply {
                            args = Arrays.stream(it.parameterTypes)
                                    .map { it.simpleName }
                                    .toArray { arrayOfNulls<String>(it) }

                            result = it.returnType.simpleName
                        }

                        genModels(models, it.returnType, visited)

                        it.name to method
                    }
                    .collect<HashMap<String, ModelMethod>>(
                            ::HashMap,
                            { m, (name, method) -> m[name] = method },
                            { obj, m -> obj.putAll(m) }
                    )
        }

        models[target.simpleName] = model
    }
}